File: experiment-14/runs/language=clojure_model=sonnet_prompt=TDD/rep1/brazilian-soccer-mcp/test/brazilian_soccer_mcp/queries_test.clj
(ns brazilian-soccer-mcp.queries-test
  (:require [clojure.test :refer :all]
            [brazilian-soccer-mcp.queries :as q]
            [brazilian-soccer-mcp.data :as data]))

(def sample-matches
  [{:home-team "Flamengo" :away-team "Palmeiras" :home-goal 2 :away-goal 1
    :season 2023 :competition "brasileirao" :round 22
    :date (java.time.LocalDate/of 2023 9 3)}
   {:home-team "Fluminense" :away-team "Flamengo" :home-goal 1 :away-goal 0
    :season 2023 :competition "brasileirao" :round 8
    :date (java.time.LocalDate/of 2023 5 28)}
   {:home-team "Palmeiras" :away-team "Corinthians" :home-goal 3 :away-goal 1
    :season 2022 :competition "brasileirao" :round 10
    :date (java.time.LocalDate/of 2022 6 15)}
   {:home-team "Flamengo" :away-team "Corinthians" :home-goal 0 :away-goal 0
    :season 2022 :competition "brasileirao" :round 5
    :date (java.time.LocalDate/of 2022 4 10)}
   {:home-team "Santos" :away-team "Flamengo" :home-goal 1 :away-goal 3
    :season 2023 :competition "copa-do-brasil" :round "semi-final"
    :date (java.time.LocalDate/of 2023 8 10)}])

(deftest find-matches-by-team-test
  (testing "finds matches where team played as home"
    (let [results (q/find-matches sample-matches {:team "Flamengo" :role :home})]
      (is (= 2 (count results)))
      (is (every? #(= "Flamengo" (:home-team %)) results))))

  (testing "finds matches where team played as away"
    (let [results (q/find-matches sample-matches {:team "Flamengo" :role :away})]
      (is (= 2 (count results)))
      (is (every? #(= "Flamengo" (:away-team %)) results))))

  (testing "finds all matches with team (home or away)"
    (let [results (q/find-matches sample-matches {:team "Flamengo"})]
      (is (= 4 (count results)))))

  (testing "finds matches for a team not present"
    (let [results (q/find-matches sample-matches {:team "Grêmio"})]
      (is (empty? results)))))

(deftest find-matches-by-season-test
  (testing "filters matches by season"
    (let [results (q/find-matches sample-matches {:season 2023})]
      (is (= 3 (count results)))
      (is (every? #(= 2023 (:season %)) results))))

  (testing "returns empty for unknown season"
    (let [results (q/find-matches sample-matches {:season 2019})]
      (is (empty? results)))))

(deftest find-matches-by-competition-test
  (testing "filters by competition"
    (let [results (q/find-matches sample-matches {:competition "brasileirao"})]
      (is (= 4 (count results)))))

  (testing "filters copa-do-brasil"
    (let [results (q/find-matches sample-matches {:competition "copa-do-brasil"})]
      (is (= 1 (count results))))))

(deftest find-matches-combined-filters-test
  (testing "filters by team AND season"
    (let [results (q/find-matches sample-matches {:team "Flamengo" :season 2023})]
      (is (= 3 (count results)))))

  (testing "filters by team AND competition"
    (let [results (q/find-matches sample-matches {:team "Flamengo" :competition "brasileirao"})]
      (is (= 3 (count results)))))

  (testing "head-to-head: finds matches between two specific teams"
    (let [results (q/find-matches sample-matches {:team1 "Flamengo" :team2 "Palmeiras"})]
      (is (= 1 (count results))))))

(deftest calculate-team-stats-test
  (testing "calculates win/draw/loss from matches"
    (let [stats (q/calculate-team-stats sample-matches "Flamengo")]
      (is (= 4 (:matches stats)))
      (is (= 2 (:wins stats)))
      (is (= 1 (:draws stats)))
      (is (= 1 (:losses stats)))
      (is (= 5 (:goals-for stats)))
      (is (= 3 (:goals-against stats)))))

  (testing "stats for team with no matches"
    (let [stats (q/calculate-team-stats sample-matches "Grêmio")]
      (is (= 0 (:matches stats)))
      (is (= 0 (:wins stats))))))

(deftest calculate-standings-test
  (testing "calculates standings sorted by points"
    (let [standings (q/calculate-standings sample-matches)]
      (is (seq standings))
      (let [first-team (first standings)]
        (is (contains? first-team :team))
        (is (contains? first-team :points))
        (is (contains? first-team :wins))
        (is (contains? first-team :draws))
        (is (contains? first-team :losses)))))

  (testing "standings are sorted by points descending"
    (let [standings (q/calculate-standings sample-matches)
          points (map :points standings)]
      (is (= points (sort > points))))))

(deftest biggest-wins-test
  (testing "returns matches sorted by goal difference descending"
    (let [wins (q/biggest-wins sample-matches 3)]
      (is (<= (count wins) 3))
      (let [diffs (map #(Math/abs (- (:home-goal %) (:away-goal %))) wins)]
        (is (= diffs (sort > diffs))))))

  (testing "returns up to n results"
    (let [wins (q/biggest-wins sample-matches 2)]
      (is (= 2 (count wins))))))

(deftest head-to-head-stats-test
  (testing "calculates head-to-head record"
    (let [stats (q/head-to-head-stats sample-matches "Flamengo" "Corinthians")]
      (is (= 1 (:total stats)))
      (is (= 0 (:team1-wins stats)))
      (is (= 1 (:draws stats)))
      (is (= 0 (:team2-wins stats))))))

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep1/tests/queries.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { loadData, type SoccerData } from '../src/loader';
import {
  filterMatches,
  teamRecord,
  headToHead,
  competitionStandings,
  filterPlayers,
  topScoringTeams,
  biggestWins,
  aggregateStats,
  brazilianPlayersByClub,
  competitionsForTeam,
} from '../src/queries';

let data: SoccerData;

beforeAll(() => {
  data = loadData();
});

describe('Feature: Data loading', () => {
  describe('Scenario: all six CSV files are loaded', () => {
    it('loads thousands of matches and players', () => {
      expect(data.matches.length).toBeGreaterThan(10000);
      expect(data.players.length).toBeGreaterThan(15000);
    });

    it('covers all three Brazilian competitions', () => {
      const comps = new Set(data.matches.map(m => m.competition));
      expect(comps.has('Brasileirão')).toBe(true);
      expect(comps.has('Copa do Brasil')).toBe(true);
      expect(comps.has('Copa Libertadores')).toBe(true);
    });

    it('loads matches from each source file', () => {
      const sources = new Set(data.matches.map(m => m.source));
      expect(sources.has('Brasileirao_Matches.csv')).toBe(true);
      expect(sources.has('Brazilian_Cup_Matches.csv')).toBe(true);
      expect(sources.has('Libertadores_Matches.csv')).toBe(true);
      expect(sources.has('BR-Football-Dataset.csv')).toBe(true);
      expect(sources.has('novo_campeonato_brasileiro.csv')).toBe(true);
    });
  });
});

describe('Feature: Match queries', () => {
  describe('Scenario: find matches between two teams', () => {
    it('returns Flamengo vs Fluminense matches in the dataset', () => {
      const matches = filterMatches(data.matches, { team: 'Flamengo', opponent: 'Fluminense' });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) {
        const involvesBoth =
          (m.homeTeam.includes('flamengo') && m.awayTeam.includes('fluminense')) ||
          (m.awayTeam.includes('flamengo') && m.homeTeam.includes('fluminense'));
        expect(involvesBoth).toBe(true);
        expect(m.date).toMatch(/^\d{4}-\d{2}-\d{2}/);
        expect(typeof m.homeGoals).toBe('number');
        expect(typeof m.awayGoals).toBe('number');
      }
    });
  });

  describe('Scenario: filter matches by team and season', () => {
    it('returns Palmeiras matches in 2019', () => {
      const matches = filterMatches(data.matches, { team: 'Palmeiras', season: 2019 });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) expect(m.season).toBe(2019);
    });
  });

  describe('Scenario: filter matches by competition', () => {
    it('returns only Copa do Brasil matches when requested', () => {
      const matches = filterMatches(data.matches, { competition: 'Copa do Brasil', limit: 50 });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) expect(m.competition).toBe('Copa do Brasil');
    });
  });

  describe('Scenario: matches are sorted by most recent first', () => {
    it('returns matches in descending date order', () => {
      const matches = filterMatches(data.matches, { team: 'Corinthians', limit: 20 });
      for (let i = 1; i < matches.length; i++) {
        expect(matches[i - 1].date >= matches[i].date).toBe(true);
      }
    });
  });
});

describe('Feature: Team statistics', () => {
  describe('Scenario: compute season record for a team', () => {
    it('reports a coherent W/D/L for Palmeiras in 2019 Brasileirão', () => {
      const rec = teamRecord(data.matches, 'Palmeiras', { season: 2019, competition: 'Brasileirão' });
      expect(rec.matches).toBeGreaterThan(0);
      expect(rec.wins + rec.draws + rec.losses).toBe(rec.matches);
      expect(rec.points).toBe(rec.wins * 3 + rec.draws);
      expect(rec.goalDifference).toBe(rec.goalsFor - rec.goalsAgainst);
    });
  });

  describe('Scenario: home vs away record', () => {
    it('sums home and away matches back to overall total', () => {
      const home = teamRecord(data.matches, 'Corinthians', { season: 2022, competition: 'Brasileirão', venue: 'home' });
      const away = teamRecord(data.matches, 'Corinthians', { season: 2022, competition: 'Brasileirão', venue: 'away' });
      const all = teamRecord(data.matches, 'Corinthians', { season: 2022, competition: 'Brasileirão' });
      expect(home.matches + away.matches).toBe(all.matches);
      expect(home.goalsFor + away.goalsFor).toBe(all.goalsFor);
    });
  });
});

describe('Feature: Head-to-head', () => {
  describe('Scenario: compare two clubs', () => {
    it('Palmeiras vs Santos returns consistent counts', () => {
      const h = headToHead(data.matches, 'Palmeiras', 'Santos');
      expect(h.matches).toBeGreaterThan(0);
      expect(h.teamAWins + h.teamBWins + h.draws).toBe(h.matches);
      expect(h.history.length).toBe(h.matches);
    });
  });
});

describe('Feature: Competition standings', () => {
  describe('Scenario: 2019 Brasileirão final table', () => {
    it('ranks Flamengo first in 2019', () => {
      const standings = competitionStandings(data.matches, 'Brasileirão', 2019);
      expect(standings.length).toBeGreaterThan(15);
      expect(standings[0].team).toBe('flamengo');
      expect(standings[0].rank).toBe(1);
      expect(standings[0].points).toBeGreaterThan(standings[1].points);
    });

    it('every standing row has matches equal to W+D+L', () => {
      const standings = competitionStandings(data.matches, 'Brasileirão', 2019);
      for (const row of standings) {
        expect(row.wins + row.draws + row.losses).toBe(row.matches);
      }
    });
  });
});

describe('Feature: Player queries', () => {
  describe('Scenario: find Brazilian players', () => {
    it('returns players with Brazil as nationality', () => {
      const players = filterPlayers(data.players, { nationality: 'Brazil', limit: 50 });
      expect(players.length).toBeGreaterThan(0);
      for (const p of players) expect((p.nationality || '').toLowerCase()).toBe('brazil');
    });

    it('returns Neymar among top-rated Brazilians', () => {
      const players = filterPlayers(data.players, { nationality: 'Brazil', sortBy: 'overall', order: 'desc', limit: 10 });
      const names = players.map(p => p.name).join(' | ').toLowerCase();
      expect(names).toContain('neymar');
    });
  });

  describe('Scenario: search by player name', () => {
    it('finds Messi by partial name', () => {
      const players = filterPlayers(data.players, { name: 'Messi', limit: 5 });
      expect(players.length).toBeGreaterThan(0);
      expect(players.some(p => p.name.toLowerCase().includes('messi'))).toBe(true);
    });
  });

  describe('Scenario: filter by club', () => {
    it('returns Santos players when filtering by club', () => {
      const players = filterPlayers(data.players, { club: 'Santos', limit: 50 });
      expect(players.length).toBeGreaterThan(0);
      for (const p of players) {
        expect((p.club || '').toLowerCase()).toContain('santos');
      }
    });
  });
});

describe('Feature: Statistical analysis', () => {
  describe('Scenario: aggregate statistics for the Brasileirão', () => {
    it('reports average goals per match in a reasonable range', () => {
      const s = aggregateStats(data.matches, { competition: 'Brasileirão' });
      expect(s.totalMatches).toBeGreaterThan(0);
      expect(s.averageGoalsPerMatch).toBeGreaterThan(1.5);
      expect(s.averageGoalsPerMatch).toBeLessThan(4);
      expect(s.homeWinRate + s.awayWinRate + s.drawRate).toBeCloseTo(1, 2);
      expect(s.homeWinRate).toBeGreaterThan(s.awayWinRate);
    });
  });

  describe('Scenario: top scoring teams', () => {
    it('returns rankings sorted descending by goals', () => {
      const rows = topScoringTeams(data.matches, { competition: 'Brasileirão', season: 2019, limit: 5 });
      expect(rows.length).toBe(5);
      for (let i = 1; i < rows.length; i++) {
        expect(rows[i - 1].goals >= rows[i].goals).toBe(true);
      }
    });
  });

  describe('Scenario: biggest wins in the dataset', () => {
    it('returns matches sorted by goal margin', () => {
      const wins = biggestWins(data.matches, { limit: 5 });
      expect(wins.length).toBe(5);
      for (let i = 1; i < wins.length; i++) {
        const a = Math.abs(wins[i - 1].homeGoals - wins[i - 1].awayGoals);
        const b = Math.abs(wins[i].homeGoals - wins[i].awayGoals);
        expect(a >= b).toBe(true);
      }
      expect(Math.abs(wins[0].homeGoals - wins[0].awayGoals)).toBeGreaterThanOrEqual(5);
    });
  });
});

describe('Feature: Cross-file relationships', () => {
  describe('Scenario: Brazilian players by club summary', () => {
    it('returns groups with non-empty clubs and positive counts', () => {
      const rows = brazilianPlayersByClub(data.players);
      expect(rows.length).toBeGreaterThan(0);
      for (const r of rows.slice(0, 10)) {
        expect(r.count).toBeGreaterThan(0);
        expect(r.club).toBeTruthy();
      }
    });
  });

  describe('Scenario: list competitions a team has played in', () => {
    it('reports Brasileirão for Palmeiras', () => {
      const comps = competitionsForTeam(data.matches, 'Palmeiras');
      expect(comps).toContain('Brasileirão');
    });
  });
});
